tests/integration/models/test_eval_methods_mass_balance.py

```diff
@@ -0,0 +1,420 @@
+                    2.80122961e-04,
+                    2.80123025e-04,
+                    2.80122961e-04,
+                    2.80123025e-04,
+                    1.59033151e-06,
+                ]
+            ),
+            None,
+        ),
+        ("mobility", 1 / 0.001, None),
+        ("normal_permeability", 1.0, None),
+        ("permeability", 1e-20, None),
+        ("porosity", 7e-3, None),
+        ("pressure_exponential", np.exp(4e-10 * 200 * pp.BAR), None),
+        (
+            "pressure_trace",
+            np.array(
+                [
+                    0.00e00,
+                    -3.00e07,
+                    0.00e00,
+                    0.00e00,
+                    -3.00e07,
+                    0.00e00,
+                    0.00e00,
+                    0.00e00,
+                    -3.00e07,
+                    -3.00e07,
+                    0.00e00,
+                    0.00e00,
+                    -3.00e07,
+                    -3.00e07,
+                    -3.00e07,
+                    -3.00e07,
+                    0.00e00,
+                    -2.48e09,
+                    0.00e00,
+                    -2.48e09,
+                    0.00e00,
+                    -2.48e09,
+                    0.00e00,
+                    -2.48e09,
+                ]
+            ),
+            None,
+        ),
+        ("reference_pressure", 0, None),
+        ("skin_factor", 0, None),
+        ("tangential_component", np.array([[1.0, 0.0]]), 0),  # check only for 0d
+        ("well_fluid_flux", 0, 2),  # use dim_restriction=2 to ignore well flux
+        ("well_flux_equation", 0, 2),  # use dim_restriction=2 to ignore well equation
+        ("well_radius", 0.1, None),
+    ],
+)
+def test_ad_operator_methods_single_phase_flow(
+    model_setup,
+    method_name: str,
+    expected_value: float | np.ndarray,
+    dimension_restriction: int | None,
+) -> None:
+    """Test that Ad operator methods return expected values.
+
+    Parameters:
+        model_setup: Prepared-for-simulation single phase flow model setup.
+        method_name: Name of the method to be tested.
+        expected_value: The expected value from the evaluation.
+        dimension_restriction: Dimension in which the method is restricted. If None,
+            all valid dimensions for the method will be considered. Can also be used
+            to test a method that is valid in all dimensions, but for the sake of
+            compactness is only tested in one dimension.
+
+    """
+    # Get the method to be tested.
+    method: Callable = getattr(model_setup, method_name)
+
+    # Obtain list of subdomain or interface grids where the method is defined.
+    domains = setup_utils.domains_from_method_name(
+        model_setup.mdg,
+        method,
+        dimension_restriction,
+    )
+
+    # Obtain operator in AD form.
+    operator = method(domains)
+
+    # Discretize (if necessary), evaluate, and retrieve numerical values.
```

Comment:
Shouldn't this be taken care of by prepare_simulation?

Reply:
Not entirely. There are two reasons this is needed: (1) the `prepare_simulation()` does not consider the non-trivial primary variable states (since these are set afterwards), and (2) some methods have a domain restriction which must be respected, i.e., via `dimension_restriction`.

Reply:
(2) should not be a problem, I think. (1) might be, I suppose.